File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/FilterEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import java.util.ArrayList;
import java.util.List;
import org.hl7.elm.r1.Filter;
import org.opencds.cqf.cql.engine.execution.State;
import org.opencds.cqf.cql.engine.execution.Variable;

public class FilterEvaluator {

    public static Object filter(Filter elm, Object source, Object condition, State state) {

        List<Object> ret = new ArrayList<>();

        if (source == null) {
            ret = null;
        }

        if (source instanceof Iterable) {

            for (Object obj : (List<?>) source) {
                try {
                    // Hmmm... This is hard without the alias.
                    // TODO: verify this works for all cases -> will scope always be present?
                    if (elm.getScope() != null) {
                        state.push(new Variable(elm.getScope()).withValue(obj));
                    }

                    if (condition instanceof Boolean && (Boolean) condition) {
                        ret.add(obj);
                    }
                } finally {
                    state.pop();
                }
            }
        }

        return ret;
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/FilterEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.hl7.elm.r1.Filter
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.execution.Variable
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlList

object FilterEvaluator {
    @JvmStatic
    fun filter(elm: Filter?, source: Value?, condition: Value?, state: State?): List? {

        if (source == null) {
            return null
        }

        val ret = mutableListOf<Value?>()

        if (source is List) {
            for (obj in source) {
                try {
                    // Hmmm... This is hard without the alias.
                    // TODO: verify this works for all cases -> will scope always be present?
                    if (elm!!.scope != null) {
                        state!!.push(Variable(elm.scope!!).withValue(obj))
                    }

                    if (condition is Boolean && condition.value) {
                        ret.add(obj)
                    }
                } finally {
                    state!!.pop()
                }
            }
        }

        return ret.toCqlList()
    }
}
